java-checks/src/main/java/org/sonar/java/checks/CollectionMethodsWithLinearComplexityCheck.java

```diff
@@ -0,0 +1,142 @@
+    builder.put(collectionContains, ImmutableSet.of(ARRAY_LIST, LINKED_LIST, COW_ARRAY_LIST, COW_ARRAY_SET, CONCURRENT_LINKED_QUEUE, CONCURRENT_LINKED_DEQUE));
+
+    MethodMatcher collectionSize = collectionMethodMatcher().name("size").withoutParameter();
+    builder.put(collectionSize, ImmutableSet.of(CONCURRENT_LINKED_QUEUE, CONCURRENT_LINKED_DEQUE));
+
+    MethodMatcher collectionAdd = collectionMethodMatcher().name("add").addParameter(TypeCriteria.anyType());
+    builder.put(collectionAdd, ImmutableSet.of(COW_ARRAY_SET, COW_ARRAY_LIST));
+
+    MethodMatcher collectionRemove = collectionMethodMatcher().name("remove").addParameter(JAVA_LANG_OBJECT);
+    builder.put(collectionRemove, ImmutableSet.of(ARRAY_LIST, COW_ARRAY_SET, COW_ARRAY_LIST));
+
+    MethodMatcher listGet = MethodMatcher.create().typeDefinition(TypeCriteria.subtypeOf("java.util.List")).name("get").addParameter("int");
+    builder.put(listGet, ImmutableSet.of(LINKED_LIST));
+    matcherActualTypeMap = builder.build();
+  }
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodInvocationTree mit = (MethodInvocationTree) tree;
+    if (!invocationInMethod(mit)) {
+      return;
+    }
+    matcherActualTypeMap.forEach((methodMatcher, actualTypes) -> {
+      if (methodMatcher.matches(mit)) {
+        Symbol target = invocationTarget(mit);
+        if (target != null && isField(target) && matchesActualType(target, actualTypes)) {
+          IdentifierTree methodName = MethodsHelper.methodName(mit);
+          reportIssue(methodName, "This call to \"" + methodName.name() + "()\" may be a performance hot spot if the collection is large.");
+        }
+      }
+    });
+  }
+
+  private static boolean invocationInMethod(MethodInvocationTree mit) {
+    Tree parent = mit.parent();
+    while (parent != null && !parent.is(Tree.Kind.METHOD)) {
+      parent = parent.parent();
+    }
+    return parent != null;
+  }
+
+  private static boolean isField(Symbol symbol) {
+    return symbol.isVariableSymbol() && symbol.owner().isTypeSymbol() && !"this".equals(symbol.name()) && !"super".equals(symbol.name());
+  }
+
+  private static boolean matchesActualType(Symbol invocationTarget, Set<String> actualTypes) {
+    return actualTypes.contains(actualType(invocationTarget).fullyQualifiedName());
+  }
+
+  @CheckForNull
+  private static Symbol invocationTarget(MethodInvocationTree mit) {
+    ExpressionTree methodSelect = mit.methodSelect();
+    if (methodSelect.is(Tree.Kind.MEMBER_SELECT)) {
+      ExpressionTree methodSelectExpression = ((MemberSelectExpressionTree) methodSelect).expression();
+      if (methodSelectExpression.is(Tree.Kind.IDENTIFIER)) {
+        return ((IdentifierTree) methodSelectExpression).symbol();
+      }
+    }
+    return null;
+  }
+
+  private static Type actualType(Symbol symbol) {
```

Comment:
I thought we agreed on the fact that we should only focus on invocation where we are 100% sure about the type. This won't work. Only looking at declarations does NOT guarantee that the actual type on which you will call a `get` or `remove` will be this one. If the variable is reassigned straight ahead, or forbidden methods are used on branches where the type has been changed, then we will raise FPs.

As a remainder, we agreed on targeting only the invocations on which the type is guaranteed. To do so, please consider types of the initializer **and all the re-assignments** that you can get from symbol usages. They all have to be of forbidden types in order to allow us to raise issues.